yarn/src/main/scala/org/apache/spark/deploy/yarn/ContainerPlacementStrategy.scala

```diff
@@ -0,0 +1,206 @@
+ * (15):
+ *
+ * requests for 4 containers with nodes: (host1, host2, host3, host4)
+ * requests for 3 containers with nodes: (host1, host2, host3)
+ * requests for 3 containers with nodes: (host1, host2)
+ *
+ * The placement ratio is 10 : 10 : 7 : 4, close to expected ratio (3 : 3 : 2 : 1)
+ *
+ * 3. If containers are existed but no matching localities, follow the method of 1 and 2.
+ *
+ * 4. If containers are existed and some localities are matched. For example if we have 1
+ * containers on each node (host1: 1, host2: 1: host3: 1, host4: 1), and the expected containers
+ * on each node would be (host1: 5, host2: 5, host3: 4, host4: 2),
+ * so the newly requested containers on each node would be updated to (host1: 4, host2: 4,
+ * host3: 3, host4: 1), 12 containers by total.
+ *
+ *   4.1 If requested container number (18) is more than newly required containers (12). Follow
+ *   method 1 with updated ratio 4 : 4 : 3 : 1.
+ *
+ *   4.2 If request container number (10) is more than newly required containers (12). Follow
+ *   method 2 with updated ratio 4 : 4 : 3 : 1.
+ *
+ * 5. If containers are existed and existing localities can fully cover the requested localities.
+ * For example if we have 5 containers on each node (host1: 5, host2: 5, host3: 5, host4: 5),
+ * which could cover the current requested localities. This algorithm will allocate all the
+ * requested containers with no localities.
+ */
+private[yarn] class LocalityPreferredContainerPlacementStrategy(
+    val sparkConf: SparkConf,
+    val yarnConf: Configuration,
+    val yarnAllocator: YarnAllocator) extends ContainerPlacementStrategy {
+
+  // Number of CPUs per task
+  private val CPUS_PER_TASK = sparkConf.getInt("spark.task.cpus", 1)
+
+  // Get the required cores of locality aware task
+  private def localityAwareTaskCores(localityAwarePendingTasks: Int): Int = {
+    localityAwarePendingTasks * CPUS_PER_TASK
+  }
+
+  // Get the expected number of locality aware containers
+  private def expectedLocalityAwareContainerNum(localityAwarePendingTasks: Int): Int = {
+    (localityAwareTaskCores(localityAwarePendingTasks) +
+      yarnAllocator.resource.getVirtualCores - 1) /
+        yarnAllocator.resource.getVirtualCores
+  }
+
+  // Update the expected locality distribution by considering the existing allocated container
+  // host distributions.
+  private def updateExpectedLocalityToCounts(localityAwarePendingTasks: Int,
+      preferredLocalityToCounts: Map[String, Int]): Map[String, Int] = {
+    val totalPreferredLocalities = preferredLocalityToCounts.values.sum
+    preferredLocalityToCounts.map { case (host, count) =>
+      val expectedCount =
+        count.toDouble * expectedLocalityAwareContainerNum(localityAwarePendingTasks) /
+          totalPreferredLocalities
+      val existedCount = yarnAllocator.allocatedHostToContainersMap.get(host)
+        .map(_.size)
+        .getOrElse(0)
+
+      if (expectedCount > existedCount) {
```

Comment:
Can this just be `(host, math.max(0, (expectedCount - existedCount).ceil.toInt))`?